cpp/src/parquet/arrow/reader_internal.cc

```diff
@@ -556,10 +556,16 @@ struct DecimalConverter<DecimalArrayType, FLBAType> {
     const int32_t byte_width =
         checked_cast<const ::arrow::FixedSizeBinaryType&>(*fixed_size_binary_array.type())
             .byte_width();
-    // allocate memory for the decimal array
-    ARROW_ASSIGN_OR_RAISE(auto data, ::arrow::AllocateBuffer(length * type_length, pool));
 
     // raw bytes that we can write to
+    std::shared_ptr<::arrow::Buffer> data;
+    // if the byte width of the FixedSizeBinaryArray is greater than or equal to the given
+    // array then we can reuse its data buffer to write the decimal array
+    if (byte_width >= type_length) {
+      data = fixed_size_binary_array.data()->buffers[1];
```

Comment:
This may change the buffer which is shared by other arrays. So I don't think this is a safe change.

Reply:
It is not shared by other arrays because it is only used here
https://github.com/apache/arrow/blob/f7947cc21bf78d67cf5ac1bf1894b5e04de1a632/cpp/src/parquet/arrow/reader_internal.cc#L743
Which is used only here
https://github.com/apache/arrow/blob/f7947cc21bf78d67cf5ac1bf1894b5e04de1a632/cpp/src/parquet/arrow/reader.cc#L488-L489
So this is function only for reading from file and we see that arrays in it are not shared